src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java

```diff
@@ -119,67 +115,44 @@ public K8sUnicastHostsProvider(Settings settings,
-            logger.trace("no pod found for selector [{}].", this.selector);
-            return cachedDiscoNodes;
-        }
-
-        for (Pod pod : pods) {
-            String status = pod.getCurrentState().getStatus();
-            logger.trace("k8s instance {} with status {} found.", pod.getId(), status);
-
-            if (!status.equals(Status.RUNNING)) {
-                logger.trace("k8s instance {} is not running - ignoring.", pod.getId());
-                continue;
-            }
-
-            try {
-                String podIp = pod.getCurrentState().getPodIP();
-
-                if (podIp.equals(currentIpAddress)) {
-                    // We found the current node.
-                    // We can ignore it in the list of DiscoveryNode
-                    logger.trace("current node found. Ignoring {} - {}", pod.getId(), podIp);
+        // get all endpoints for service
+        final Endpoints endpoints = getKubernetes().endpointsForService(this.selector, null);
+
+        int podsCount = 0;
+        // populate discovery nodes list
+        for (EndpointSubset es : endpoints.getSubsets()) {
+            for (EndpointAddress address : es.getAddresses()) {
+                podsCount++;
+                final String endpointIp = address.getIP();
+                if (endpointIp.equals(currentIpAddress)) {
+                    logger.trace("Current node found. Ignoring {}", endpointIp);
                 } else {
-                    List<Container> containers = KubernetesHelper.getContainers(pod);
-                    for (Container container : containers) {
-                        logger.trace("pod " + pod.getId() + " container: " + container.getName() + " image: " + container.getImage());
-                        List<Port> ports = container.getPorts();
-                        for (Port port : ports) {
-                            Integer containerPort = port.getContainerPort();
-                            if (containerPort == 9300) {
-                                String address = podIp.concat(":").concat(containerPort.toString());
-
-                                // pod IP is a single IP Address. We need to build a TransportAddress from it
-                                TransportAddress[] addresses = transportService.addressesFromString(address);
-
-                                logger.trace("adding {}, address {}, transport_address {}, status {}", pod.getId(),
-                                        podIp, addresses[0], status);
-                                cachedDiscoNodes.add(new DiscoveryNode("#cloud-" + pod.getId() + "-" + 0, addresses[0], Version.CURRENT));
-                            }
-                        }
+                    // pod IP is a single IP Address. We need to build a TransportAddress from it
+                    final TransportAddress[] addresses;
+                    try {
+                        addresses = transportService.addressesFromString(endpointIp.concat(":9300"));
```

Comment:
While I'm hardcoding port 9300 and thus limiting the user choices here, I'm concerned that a multiport service will make it harder for us to determine which port to use.